crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1655,37 +1650,68 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = setcomp;
 
         self.infer_expression(elt);
-        for comprehension in generators {
-            self.infer_comprehension(comprehension);
-        }
+        self.infer_comprehensions_scope(generators);
     }
 
-    fn infer_comprehension(&mut self, comprehension: &ast::Comprehension) {
-        self.infer_definition(comprehension);
-        for expr in &comprehension.ifs {
-            self.infer_expression(expr);
+    fn infer_comprehensions_scope(&mut self, comprehensions: &[ast::Comprehension]) {
```

Comment:
We've inherited some awfully confusing naming from the CPython grammar and AST :( If we have `[y for x in [] for y in x]`, that's a single "list comprehension", whose AST node has a `generators` attribute containing two... Comprehensions, `for x in []` and `for y in x`. So not only is the term "generator" overloaded here from its usual meaning, but we're also using the term "comprehension" itself to mean two entirely different things, one of which is part of the other 🤯

(To be clear, this is not a critique of this PR, just a lament for the state of naming in this part of the AST. We could fix this in our AST, but then we'd be inventing our own naming scheme that doesn't match the CPython AST, and that doesn't seem great either.)

One way to maybe clarify this naming a bit would be to switch from `list_comprehension`, `dict_comprehension`, etc all through these method names to instead use `listcomp`, `dictcomp`, etc when we are referring to the outer expression, and reserve the full word "comprehension" for the thing actually named `Comprehension` in the AST, which is just the inner `for x in y` part. I wouldn't exactly call that _clear_, but given what we're working with, it might at least be better? It at least matches the names used in the AST and avoids using the word "comprehension" to mean two different things.


In any case, I don't like the names `infer_comprehensions_scope` and `infer_comprehension_scope` for these two methods. The individual "Comprehension" nodes inside a list/set/dict/gen-comp are not different scopes, so the use of "scope" here is misleading.

If we go with my suggestion above to rename e.g. `infer_list_comprehension_expression` to `infer_listcomp_expression`, then I think we could name these two just `infer_comprehensions` and `infer_comprehension`.

Reply:
I agree with the sentiment, thanks for sharing that.

> (To be clear, this is not a critique of this PR, just a lament for the state of naming in this part of the AST. We could fix this in our AST, but then we'd be inventing our own naming scheme that doesn't match the CPython AST, and that doesn't seem great either.)

I think, if we want, we _can_ change the names in our AST as we've done so in the past as well (https://github.com/astral-sh/ruff/pull/8064, https://github.com/astral-sh/ruff/pull/6379, https://github.com/astral-sh/ruff/pull/6253, etc.). There are other recommendations here: https://github.com/astral-sh/ruff/discussions/6183

> One way to maybe clarify this naming a bit would be to switch from `list_comprehension`, `dict_comprehension`, etc all through these method names to instead use `listcomp`, `dictcomp`, etc when we are referring to the outer expression, and reserve the full word "comprehension" for the thing actually named `Comprehension` in the AST, which is just the inner `for x in y` part.

From my perspective, both `listcomp` and `list_comprehension` seems equivalent as the former seems like a short version of the latter. Is the reason for recommending `listcomp` because of it's presence in the CPython AST?

> In any case, I don't like the names `infer_comprehensions_scope` and `infer_comprehension_scope` for these two methods. The individual "Comprehension" nodes inside a list/set/dict/gen-comp are not different scopes, so the use of "scope" here is misleading.

Makes sense. I can change this.

Reply:
I've not renamed the `list_comprehension` to `listcomp` as I feel like the "expression" prefix in `list_comprehension_expression` gives a signal that this is different than a simple "comprehension". Regardless, I don't have a strong opinion, so I'm happy to rename it if you think `listcomp` is better.

Reply:
Yeah, mostly what my suggestion was aiming for was to hew very closely to the AST naming, in hopes that this would at least give some guideposts to readers. But I'm OK with your approach and relying on `_expression` to distinguish.